TensorLib/Basic.lean

```diff
@@ -2,4 +2,98 @@ namespace TensorLib
 
 def hello := "world"
 
+class TensorElement (a: Type) where
+  add: a -> a -> a
+  -- etc. enough to implement numpy ops
+
+inductive TensorData (a: Type) [TensorElement a] where
+| td_root (x: List a)
+| td_node (tl: List (TensorData a))
+
+-- Awkward... I'm sure this exists somewhere in the stdlib
+def sequence {a: Type} (x: List (Option a)): Option (List a) :=
+  match x with
+  | .nil => some .nil
+  | .cons none _ => none
+  | .cons (some y) ys =>
+    match sequence ys with
+    | none => none
+    | some ys' => some (.cons y ys')
+
+-- somehow prove termination here
+def Shape {a: Type} [TensorElement a] (x: TensorData a): Option (List Nat) :=
+  match x with
+  | TensorData.td_root x => some [List.length x]
+  | TensorData.td_node xs =>
+    match sequence (.map Shape xs) with
+    | none => none
+    | some .nil => some .nil
+    | some (.cons x xs) =>
+      if List.all xs (fun x' => x == x') then
+        some (List.cons (1 + List.length xs) x)
+      else
+        none
+
```

Comment:
We may want to pack a TensorData and a proof that Shape ... = some ... in a dependent record.

Reply:
We'll work this out in the next PR.